src/NHibernate/Type/OneToOneType.cs

```diff
@@ -59,12 +59,31 @@ public override bool IsOneToOne
 
 		public override bool IsDirty(object old, object current, ISessionImplementor session)
```

Comment:
Because this code is not identical for `OneToOneType` and `ManyToOneType` it would be better to move it to `EntityType` (the base of both types). 

For discussion: 
- do we want to implement `IsDirty(object old, object current, ISessionImplementor session)` in `EntityType` or;
- create a new method `IsDirtyCore...` and call it from overrides? /cc @bahusoid @fredericDelaporte @maca88

Reply:
The implementation of IsDirty in this case is identical to the implementation of the ManyToOneType. Maybe the common functionality can be combined in common ToOneType in a future enhancement. The purpose of this change set is only to correct the above mentioned issue.

Reply:
@hazzik  Either way works for me. Simply moving IsDirty to base looks as less code...

It seems `IsModified` should also be modified accordingly for `OneToOneType`. But have no idea when it's really needed...

Reply:
@bahusoid `IsModified` is inherited from `AbstractType` which returns true/false based on `IsDirty`.

Reply:
I made a typo. It should be `code is now identical` not `code is no*t* identical`. Fixed.

Reply:
> IsModified is inherited from AbstractType which returns true/false based on IsDirty

And still for some reasons `ManyToOne` overrides it:
https://github.com/nhibernate/nhibernate-core/blob/18667cfbdbbb0ee7743be4e1725583052d38d007/src/NHibernate/Type/ManyToOneType.cs#L137-L151

So if it's required here it seems it's now required for OneToOne too

Reply:
I think the difference here might be due to the fact the owner of a many-to-one can be modified whereas the one-to-one cannot.

Reply:
I agree there _might_ be an issue here, but it is unrelated to the one that this pull request attempts to fix. I.E. It looks like `IsModified` _may_ not handle the case when `old` is an instance of a OneToOneType and not an identifier. If you can verify this is the case, then I'd suggest creating a separate issue to correct this. After all, this problem would exist despite these changes.

Reply:
> IsModified is inherited from AbstractType which returns true/false based on IsDirty.

Actually this statement is wrong. `IsModified` is overridden in `OneToOneType` and always returns false:
https://github.com/nhibernate/nhibernate-core/blob/eb137cc8668982ae0c0f68bd264e834c138324f8/src/NHibernate/Type/OneToOneType.cs#L70-L73

Reply:
> IsModified may not handle the case when old is an instance of a OneToOneType and not an identifier

I don't get why you think it's unrelated... It's related.
Old logic: handles both identifier and entity type in old parameter (because it simply returns null)
If you make IsModified call IsDirty - your new logic will fail when `IsModified` is called with identifier. Because now `GetIdentifier` might be called on `old` instance.

Reply:
You are correct, `IsModified` always returns `false`. This is consistent with the original behavior of the type that you cannot modify owner of the owner of the OneToOneType once assigned. I'm not against changing this behavior but I think that doing so would be the topic of another issue. `IsDirty` determines whether or not the parent should be considered dirty or not, while `IsModified` determines whether or not an update is required to the identifier of the OneToOneType. Since updates are not currently permitted, `IsModified` always returns `false`.

Reply:
Maybe I just don't understand something. But from what I see [DefaultFlushEntityEventListener.DirtyCheck](https://github.com/nhibernate/nhibernate-core/blob/6ab71d08b6c7f0a9c95895e94700b466b0743d6e/src/NHibernate/Event/Default/DefaultFlushEntityEventListener.cs#L426-L496) can call both methods for obtaining dirty properties depending on current session state:
https://github.com/nhibernate/nhibernate-core/blob/6ab71d08b6c7f0a9c95895e94700b466b0743d6e/src/NHibernate/Event/Default/DefaultFlushEntityEventListener.cs#L452

https://github.com/nhibernate/nhibernate-core/blob/6ab71d08b6c7f0a9c95895e94700b466b0743d6e/src/NHibernate/Event/Default/DefaultFlushEntityEventListener.cs#L480

With only difference that `IsModified` is called on databaseSnapshot (and identifier related code was added to many-to-one to handle such db snapshots). So to me it looks like those method should behave the same as both of them can be called for dirty check.

So regarding
>IsDirty determines whether or not the parent should be considered dirty or not, while IsModified determines whether or not an update is required to the identifier of the OneToOneType

Can you clarify where did you get this difference and how it plays with `DefaultFlushEntityEventListener.DirtyCheck`? And I don't really understand what `determines whether or not an update is required to the identifier of the OneToOneType` means considering identifier of the entity is not updateable. And if you mean in parent object - it's still can be changed from null to something and vice-versa.

Reply:
`IsDirty` has the following comment:
```C#
/// <summary>
/// Should the parent be considered dirty, given both the old and current 
/// field or element value?
/// </summary>
```

`IsModified` has the following comment
```C#
/// <summary>
/// Says whether the value has been modified
/// </summary>
```
Pseudo code for `DefaultFlushEntityEventListener.DirtyCheck` the following:

L428-434 Get entity, current property values, loaded property values, id, and persiseter.
L436 Check if the Interceptor handled the Dirty check.
L443 If the Interceptor did not handle the dirty check, then attempt to dirty check the entity.
L448 If the loaded state of the entity is null, then the entity is transient and cannot be dirty checked.
L449-453 If the entity was loaded then compare current property values against loaded property values to determine if the entity is dirty.
L454 If the entity was deleted, compare the current property values against the deleted state.
L474-484 Get the current values from the database if select before update is required or check the id against the unsaved value. If a database entry exists or the id is not equal to the unsaved value, check for modified property values.

Based on the above and the fact that the OneToOneType is not updatable, I see no reason why `IsModified` should ever return anything but `false` for the OneToOneType.

Reply:
> L448 If the loaded state of the entity is null, then the entity is transient and cannot be dirty checked.

`loadedState` can also be null if entity is added to session via via `session.Update`. It actually has explicit comment about it... So I expect with current PR state we still have issue with cache when owner is added to session via  `session.Update` and then updated. Can you prove me wrong with test case?